### ISSUE TITLE
Implement MediaSession API to support voice interactions for video/media playback

### ISSUE THREAD

**Issue by mcomella** [mozilla-mobile/firefox-tv#930]
## User benefit
On the new Fire TV cube device, you can interact with video/media using voice and [the MediaSession API](https://developer.amazon.com/blogs/appstore/post/623addd5-6184-4e7f-a969-123a653cfb1a/implementing-voice-control-with-the-media-session-api-on-amazon-fire-tv). This would be convenient for users.

## Requirements
- Investigate what we'd have to do to integrate
- Integrate if possible

**Comment by mcomella** [comment]
An additional link: https://developer.amazon.com/docs/fire-tv/mediasession-api-integration.html

**Comment by mcomella** [comment]
Looking briefly into this, it looks like we need to instantiate a `MediaSessionCompat` instance and add some callbacks. However, WebView doesn’t appear to have APIs to query/modify audio/video playback so we might need to do this through JavaScript.

**Comment by mcomella** [comment]
fwiw, there is a [`MediaSession` JS API](https://developer.mozilla.org/en-US/docs/Web/API/Media_Session_API) which makes this a little harder to google. :)

**Comment by mcomella** [comment]
The code we write could be usable by others trying to integrate `MediaSession` with `WebView`: we should consider making it a library.

**Comment by mcomella** [comment]
I'd estimate this is a size M though it could turn into a size L if the JS turns out to be difficult/fragile.

**Comment by mcomella** [comment]
I think this is as straight-forward as creating a `MediaSession` and registering callbacks. However, one unclear thing: when do we need to call `release`? The docs seem to assume that media Activities will playback for their whole lifecycle (thus in `onDestroy`) but videos can be created on each page: should we create a new session for each page? Here's the code I wrote in onCreate to test without a device:
```kotlin
        mediaSession = MediaSessionCompat(this, "lol")
        val pb = PlaybackStateCompat.Builder()
                .setActions(PlaybackStateCompat.ACTION_PLAY or
                        PlaybackStateCompat.ACTION_PAUSE)
                .build()

        mediaSession.setPlaybackState(pb)
        mediaSession.setCallback(object : MediaSessionCompat.Callback() {
            val browserFragment get() = supportFragmentManager.findFragmentByTag(BrowserFragment.FRAGMENT_TAG) as BrowserFragment?
            override fun onPlay() {
                Log.d("lol", "play called")
                (browserFragment?.webView as FirefoxAmazonWebView?)?.evalJS("var vid = document.getElementsByTagName('video')[0]; vid.play();")
            }

            override fun onStop() {
                Log.d("lol", "pause called")
                (browserFragment?.webView as FirefoxAmazonWebView?)?.evalJS("var vid = document.getElementsByTagName('video')[0]; vid.pause();")
            }
        })

        val contr = MediaControllerCompat(this, mediaSession)
        MediaControllerCompat.setMediaController(this, contr)

        // Test code to call callbacks
        contr.transportControls.play()
```

**Comment by mcomella** [comment]
Related notes:
- The remote media buttons work for free on youtube.com/tv but not on other sites: I assume youtube handles the button events themselves.
- Before this implementation, "Alexa, play" doesn't seem to work consistently on youtube.com/tv (which handles media buttons) and doesn't work at all on other sites. I'm guessing Alexa is, by default, sending media button events when these commands are spoken.

"Other sites" is being tested with this site: https://www.w3.org/2010/05/video/mediaevents.html

**Comment by mcomella** [comment]
> "Alexa, play" doesn't seem to work consistently on youtube.com/tv (which handles media buttons)

I filed #936 for this issue. "Alexa pause" seems to send media button events, it seems like we'd only need to implement this bug to get media playback on pages that don't support media button events themselves. This is #935 for the hardware remote media buttons (which would be using the same code to find videos on the page and stop them).

**Comment by mcomella** [comment]
To summarize:
- If we fix #936, we'll fix "Alexa pause/play" functionality on youtube.com/tv (but not other MediaSession functionality like seeking, next/previous, restart).
- This bug will create voice support for seek, next/previous, and restart. We can decide if we want to do this only for youtube (it's simpler to code specifically to its interface instead of the general case) or for all sites.
- #935 is media button support for non-youtube sites (i.e. ones that don't support media buttons themselves): it will share code with this bug, if we support non-youtube sites.

@Sdaswani, what are we trying to accomplish with this bug? For voice, do we want to support:
1) Just play/pause on youtube
2) Play/pause/seek/prev-next/restart on youtube
3) ^ on all sites

**Comment by mcomella** [comment]
Alexa, without MediaSession, is sending media key events: https://github.com/mozilla-mobile/firefox-tv/issues/936#issuecomment-398216178

**Comment by mcomella** [comment]
I managed to get MediaSession callbacks to start working: we had to call `requestAudioFocus`. This means we'll have to figure out all the details of what calling this method, and thus being a "media app", entails.

**Comment by mcomella** [comment]
> I managed to get MediaSession callbacks to start working: we had to call `requestAudioFocus`

And we have to call:
```kotlin
val mediaController = MediaControllerCompat(this, mediaSessionCompat)
        MediaControllerCompat.setMediaController(this, mediaController)
```

**Comment by mcomella** [comment]
Next steps:
- Do we want to focus on both audio and video or just video?
- Figure out requirements to comply with being a media app. Google has video app [doc1](https://developer.android.com/guide/topics/media-apps/video-app/building-a-video-player-activity) & [doc2](https://developer.android.com/guide/topics/media-apps/video-app/mediasession-callbacks) and [general doc](https://developer.android.com/guide/topics/media-apps/audio-focus). Amazon has [doc1](https://developer.amazon.com/docs/fire-tv/multimedia-app-requirements.html#requirement-14-avoid-using-ondestroy) and [doc2](https://developer.amazon.com/docs/fire-tv/managing-audio-focus.html).
- Add MediaSession init code
- requestAudioFocus at appropriate times and handle other media app requirements
- Add JS code to modify playback state

**Comment by mcomella** [comment]
Notes on being a good media app:
- "For video apps, your app should pause playback when voice capabilities are invoked (in your OnPause() method). When your app resumes, playback should remain paused until the user presses the PLAY button on the remote." (via [here](https://developer.amazon.com/docs/fire-tv/managing-audio-focus.html#audiofocusvoice)) This isn't what the [Media Session Sample App](https://www.amazon.com/gp/product/B07CMDMMP9) does.

**Comment by mcomella** [comment]
@Sdaswani This is going to take longer than expected: I can't get the MediaSession API to work consistently. It's hard to make accurate estimates because I don't fully understand what's required to implement the API.

My current code [is here.](https://github.com/mcomella/mediasession-with-alexa-testing/blob/4f908cfa6a6be169ab6eaa2cb496bbcaff1c7deb/app/src/main/java/xyz/mcomella/mediasession/MainActivity.kt) When I restart the device (i.e. get a clean slate), my code does not work. However, if I open [the Media Session Sample app](https://www.amazon.com/gp/product/B07CMDMMP9) first and then open my app, it works correctly – I'm guessing I'm not correctly managing the audio focus state and it's making my app interact with the system strangely.

I plan to dig into the Google sample code next to see if I can figure out what I'm missing.

**Comment by mcomella** [comment]
> I can't get the MediaSession API to work consistently.

The Google sample app, [Universal Media Player](https://github.com/googlesamples/android-UniversalMusicPlayer), does not work in the same ways that my code does not work.

This makes me wonder if I should just go full steam ahead with my code, that sort of works, and figure out the edge case later.

**Comment by mcomella** [comment]
Another issue I found: when I rebooted the device, opened the sample app (paused twice to get my app to work), opened my app to test and specified, "Alexa pause", my app didn't appear to receive the media events but the google sample app did, which doesn't really make sense to me (maybe they can register as a system provider that takes media events before the app is even opened?). But it's strange because I was granted audio focus (but "Alexa pause" causes me to lose it).

**Comment by mcomella** [comment]
Another weird experience: if I start the app, then go to a website with HTML video, and say, "Alexa play", nothing happens. However, if I start on the homepage, say, "Alexa play", the media command will be received. If I then go to the page with HTML video and say "Alexa play", the media command will be received.

I wonder if the WebView is managing audio focus somehow (in the former case, if I play the video, I will lose audio focus).

---

To summarize, to get my code to work, I need to (after a restart):
- Open the Media Session Sample App, say "Alexa pause" twice (first time won't work)
- Open Firefox, say "Alexa play" before web content

Media commands should now work.

**Comment by mcomella** [comment]
Actually, with `mediaSession.isActive = true`, I no longer need to launch the Media Session Sample App before voice commands start working (though I think the first one might be delayed).

**Comment by mcomella** [comment]
Okay, I seem to have figured out why I've been getting inconsistent behavior between `isActive` and this part:
```kotlin
        val pb = PlaybackStateCompat.Builder()
                .setActions(SUPPORTED_ACTIONS)
                .setState(PlaybackStateCompat.STATE_PLAYING, PlaybackStateCompat.PLAYBACK_POSITION_UNKNOWN, 1.0f)
                .build()
```

On an fresh run, the voice commands work correctly. However, if I set the state to `PAUSED` instead, Alexa states "What would you like to hear?" instead of executing the play command, which was unexpected to me. This brings up the question: should Alexa be able to start a video that was not playing through voice?

**Comment by mcomella** [comment]
@aminalhazwani "Alexa next/previous" will dispatch a "Media next/previous" keyboard button event. Some websites may handle this, e.g. youtube will advance to the next video, while others may not. When websites handle it, it's a good experience. However, when they don't, the user is left waiting for an action that will never happen. We can't tell when a website will handle the event but should we display a toast like, "Received next command" (with better copy), to notify the user their action has been received?

**Comment by mcomella** [comment]
Status update: I can get the device to acknowledge I want to include FF/rewind/next/previous commands but they're not getting delivered to my MediaSession: I wonder if it's because I'm not updating the playback position regularly.

**Comment by mcomella** [comment]
Status update: I've opened a PR for a basic implementation that does not conform to the MediaSession APIs. I filed issues for follow-ups to better conform, which we can prioritize.

**Comment by HiralSModi** [comment]
We tested the 2.2 APK on FireTV Cube. 
Steps:

1. "Alexa, open Youtube" - This will launch the YouTube.com app.
2. Select  Firefox. (version 2.2)
3. Then, select a video to play and click on it to start playback.
4. "Alexa, pause" - video paused
5. Again say "Alexa pause" 
Expect Behavior: Nothing happens
Observed Behavior: Video resumes
Summary: It doesn't take utterances accurately and toggles between pause and play.
Sometimes FF was crashing while doing pause and play.

**Comment by HiralSModi** [comment]
[crash_log_cube.txt](https://github.com/mozilla-mobile/firefox-tv/files/2178912/crash_log_cube.txt)
Crash Log

**Comment by Sdaswani** [comment]
Thanks @HiralSModi I filed some new tickets: https://github.com/mozilla-mobile/firefox-tv/issues/965 https://github.com/mozilla-mobile/firefox-tv/issues/966

**Comment by aminalhazwani** [comment]
> We can't tell when a website will handle the event but should we display a toast like, "Received next command" (with better copy), to notify the user their action has been received?

@mcomella sorry for the late reply, I missed the notification. Rather than a visual feedback we could opt for instant voice feedback, something like "Ok" or "Got it". If possible we can then set a small timeout and the let users know that the website is not responding to the voice commands.

**Comment by aminalhazwani** [comment]
> 4. "Alexa, pause" - video paused
> 5. Again say "Alexa pause"
  Expect Behavior: Nothing happens

@HiralSModi rather than "Nothing happens" what if Alexa says something like "Video/Media is paused", or "Video is already paused"? So on first interaction we have a visual/audio feedback coming from the video/audio itself. On second interaction we have Alexa jumping into the conversation for better clarity.
